File: services/rails-svc/app/controllers/business_controller.rb
class BusinessController < ApplicationController
  # GET /api/external/mock — the inert target the HTTP faults call back into.
  def mock
    delay = params[:delayMs].to_i
    sleep(delay / 1000.0) if delay.positive?
    render json: { ok: true, seq: params[:seq].to_i, op: params[:op].to_i, delayMs: delay }
  end

  # GET /api/dispatch/:channel — routed from "business#dispatch_channel"
  # (the action is NOT named `dispatch`: that collides with
  # ActionController::Metal#dispatch and would break request routing).
  def dispatch_channel
    channel = params[:channel]
    return render(json: { error: "unknown channel" }, status: :not_found) unless CHANNELS.include?(channel)

    delay = params[:delayMs].to_i
    sleep(delay / 1000.0) if delay.positive?
    render json: { channel: channel, dispatched: true, delayMs: delay }
  end

  # GET /api/payments/history — a real ActiveRecord read (ORM scope).
  def payments_history
    customer_id = (params[:customerId] || 1).to_i
    limit = [[(params[:limit] || 10).to_i, 1].max, 100].min
    rows = Payment.where(customer_id: customer_id).order(:id).limit(limit)
                  .pluck(:id, :order_id, :customer_id, :amount_cents, :status)
    render json: rows
  end
end

File: services/symfony-svc/overlay/src/Controller/BusinessController.php
<?php

namespace App\Controller;

use App\Entity\Payment;
use Doctrine\ORM\EntityManagerInterface;
use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\HttpFoundation\JsonResponse;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\Routing\Attribute\Route;

class BusinessController extends AbstractController
{
    use Common;

    public function __construct(private EntityManagerInterface $em)
    {
    }

    // GET /api/external/mock: the inert target the HTTP faults call back into.
    #[Route('/api/external/mock', methods: ['GET'])]
    public function mock(Request $r): JsonResponse
    {
        $delay = $this->intParam($r, 'delayMs', 0);
        if ($delay > 0) {
            usleep($delay * 1000);
        }
        return new JsonResponse([
            'ok' => true,
            'seq' => $this->intParam($r, 'seq', 0),
            'op' => $this->intParam($r, 'op', 0),
            'delayMs' => $delay,
        ]);
    }

    // GET /api/dispatch/{channel}
    #[Route('/api/dispatch/{channel}', methods: ['GET'])]
    public function dispatch(Request $r, string $channel): JsonResponse
    {
        if (! in_array($channel, self::CHANNELS, true)) {
            return new JsonResponse(['error' => 'unknown channel'], 404);
        }
        $delay = $this->intParam($r, 'delayMs', 0);
        if ($delay > 0) {
            usleep($delay * 1000);
        }
        return new JsonResponse(['channel' => $channel, 'dispatched' => true, 'delayMs' => $delay]);
    }

    // GET /api/payments/history: a real Doctrine read (php.doctrine + php.pdo).
    // ?hold=1 briefly holds the connection so the pool-saturation fan-out produces
    // genuinely overlapping DB spans across workers.
    #[Route('/api/payments/history', methods: ['GET'])]
    public function paymentsHistory(Request $r): JsonResponse
    {
        $customerId = $this->intParam($r, 'customerId', 1);
        $limit = max(1, min($this->intParam($r, 'limit', 10), 100));
        if ($this->intParam($r, 'hold', 0) > 0) {
            $this->em->getConnection()->executeQuery('SELECT pg_sleep(?)', [0.3]);
        }
        $rows = $this->em->getRepository(Payment::class)
            ->findBy(['customerId' => $customerId], ['id' => 'ASC'], $limit);
        $out = array_map(static fn (Payment $p) => [
            'id' => $p->id,
            'order_id' => $p->orderId,
            'customer_id' => $p->customerId,
            'amount_cents' => $p->amountCents,
            'status' => $p->status,
        ], $rows);
        return new JsonResponse($out);
    }
}
